src/data_types/indexed_tensor.hpp

```diff
@@ -0,0 +1,230 @@
+// SPDX-License-Identifier: MIT
+/**
+ * @file indexed_tensor.hpp
+ * File providing functions to carry out tensor calculus operations via index patterns.
+ */
+#pragma once
+
+#include "tensor.hpp"
+#include "tensor_index_tools.hpp"
+#include "vector_index_tools.hpp"
+
+namespace tensor_tools {
+
+/// A boolean, true if the type is an IndexedTensor, false otherwise.
+template <class T>
+inline constexpr bool enable_indexed_tensor = false;
+
+/**
+ * @brief A class to capture the description of a tensor indexed at a specific component.
+ * This class should not be explicitly declared in user code. It is the output of a call
+ * to the index<...> function and is an input to the tensor_mul function.
+ *
+ * @tparam TensorType The tensor being indexed.
+ * @tparam TypeSeqVectorIndexIdMap A TypeSeq containing VectorIndexIdMaps describing the
+ *              indices used to access the tensor.
+ */
+template <class TensorType, class TypeSeqVectorIndexIdMap>
+class IndexedTensor
+{
+public:
+    /// The TensorIndexIdMap describing how the tensor is indexed.
+    using index_pattern = TypeSeqVectorIndexIdMap;
+    /// The type of the tensor being indexed.
+    using tensor_type = TensorType;
+
+    static_assert(
+            type_seq_has_unique_elements_v<index_pattern>,
+            "You should not have more than two of any one index in an index expression. "
+            "Additionally repeated indices should not be associated with two covariant or two "
+            "contravariant indices.");
+
+private:
+    TensorType& m_tensor;
+
+public:
+    /**
+     * @brief A constructor of an indexed tensor.
+     * @param tensor The tensor to be indexed.
+     */
+    explicit KOKKOS_FUNCTION IndexedTensor(TensorType& tensor) : m_tensor(tensor) {}
+
+    IndexedTensor(IndexedTensor const&) = delete;
+
+    IndexedTensor(IndexedTensor&&) = delete;
+
+    IndexedTensor& operator=(IndexedTensor const&) = delete;
```

Comment:
```suggestion
    IndexedTensor(IndexedTensor&&) = delete;
    
    KOKKOS_DEFAULTED_FUNCTION ~IndexedTensor() noexcept = default;

    IndexedTensor& operator=(IndexedTensor const&) = delete;
```

The fifth one is the destructor actually https://en.cppreference.com/w/cpp/language/rule_of_three